apache/metron
# METRON-1643: Create a REGEX_ROUTING field transformation

## Contributor Comments
A relatively common pattern is to choose the value of a field based on if another field matches a set of regexes.  This can be done via stellar with match, but if the list of possible regexes are large, that can be unwieldy.  To ease that burden, we should have a REGEX_ROUTING field transformation.

## Manual Test Script
Ensure that `ZOOKEEPER`, `BROKERLIST` and `METRON_HOME` are set.
* Create a `dummy` topic
```
/usr/hdp/current/kafka-broker/bin/kafka-topics.sh --zookeeper $ZOOKEEPER --create --topic dummy --partitions 1 --replication-factor 1
```
* Open $METRON_HOME/config/zookeeper/parsers/dummy.json and save
```
{
  "parserClassName" : "org.apache.metron.parsers.csv.CSVParser"
  ,"sensorTopic" : "dummy"
  , "parserConfig": {
     "columns" : { "name": 0
                 , "first_field": 1
                 , "timestamp" : 2
                 }
   }
  ,"fieldTransformations" : [
    {
     "transformation" : "REGEX_SELECT"
    ,"input" :  "name"
    ,"output" :  "is_metron"
    ,"config" : {
      "true" : "^metron.*"
                }
    }
                           ]

}
```
*  Push configs via `$METRON_HOME/bin/zk_load_configs.sh --mode PUSH -z $ZOOKEEPER -i $METRON_HOME/config/zookeeper/`
* Stop the parsers in ambari
* Start the `dummy` parser
```
$METRON_HOME/bin/start_parser_topology.sh -k $BROKERLIST -z $ZOOKEEPER -s dummy
```
* Open `~/data.json` and save:
```
notmetron,f1,1
metron,f2,2
metron,f3,3
metron,f4,4
metron,f5,5
metron,f6,6
metron,f7,7
notmetroneither,f8,8
```
* Send data into the `dummy` topic via
```
cat data.json | /usr/hdp/current/kafka-broker/bin/kafka-console-producer.sh --broker-list $BROKERLIST --topic dummy
```
You should examine the `dummy` index in elasticsearch and ensure that messages with the `name` field of `metron` have a field called `is_metron` set to `true`.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  


In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:


### For code changes:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```


### For documentation related changes:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.
